lib/improver/tests/utilities/test_cube_extraction.py

```diff
@@ -0,0 +1,165 @@
+    ycoord = DimCoord(np.arange(3), "projection_y_coordinate")
+    xcoord = DimCoord(np.arange(3), "projection_x_coordinate")
+
+    cube = iris.cube.Cube(data, long_name="probability_of_precipitation",
+                          dim_coords_and_dims=[(threshold, 0), (ycoord, 1),
+                                               (xcoord, 2)], units="1")
+    return cube
+
+
+class Test_parse_constraint_list(IrisTest):
+
+    """ Test function to parse constraints and units into dictionaries """
+
+    def setUp(self):
+        """ Set up some constraints to parse """
+        self.constraints = ["percentile=10", "threshold=0.1"]
+
+    def test_basic_no_units(self):
+        """ Test simple key-value splitting with no units """
+        cdict, udict = parse_constraint_list(self.constraints, None)
+        self.assertEqual(cdict["percentile"], 10)
+        self.assertEqual(cdict["threshold"], 0.1)
+        self.assertFalse(udict)
+
+    def test_some_units(self):
+        """ Test units list containing "None" elements is correctly parsed """
+        units = ["None", "mm h-1"]
+        _, udict = parse_constraint_list(self.constraints, units)
+        self.assertEqual(udict["threshold"], "mm h-1")
+        self.assertNotIn("percentile", udict.keys())
+
+    def test_unmatched_units(self):
+        """ Test for ValueError if units list does not match constraints """
+        units = ["mm h-1"]
+        msg = "units list must match constraints"
+        with self.assertRaisesRegexp(ValueError, msg):
+            parse_constraint_list(self.constraints, units)
+
+
+class Test_extract_subcube(IrisTest):
+
+    """ Test function to extract subcube according to constraints """
+
+    def setUp(self):
+        """ Save temporary input cube to file """
+        self.precip_cube = set_up_precip_probability_cube()
+        self.directory = mkdtemp()
+        self.filepath = os.path.join(self.directory, 'tmp_precip_cube.nc')
+        iris.save(self.precip_cube, self.filepath)
+
+    def tearDown(self):
+        """ Remove temporary path created for testing """
+        call(['rm', '-f', self.filepath])
+        call(['rmdir', self.directory])
+
+    def test_basic_no_units(self):
+        """ Test cube extraction for single constraint without units """
+        constraint_dict = {"name": "probability_of_precipitation"}
+        cube = extract_subcube(self.filepath, constraint_dict, None)
+        self.assertIsInstance(cube, iris.cube.Cube)
+        reference_data = self.precip_cube.data
+        self.assertArrayEqual(cube.data, reference_data)
+
+    def test_basic_with_units(self):
+        """ Test cube extraction for single constraint with units """
+        constraint_dict = {"threshold": 0.1}
+        units_dict = {"threshold": "mm h-1"}
+        cube = extract_subcube(self.filepath, constraint_dict, units_dict)
+        self.assertIsInstance(cube, iris.cube.Cube)
+        self.assertEqual(cube.coord("threshold").units, "mm h-1")
```

Comment:
Will become a test that the units have been converted back.

Reply:
Done